app/client/components/RegionFocusSwitcher.ts

```diff
@@ -0,0 +1,684 @@
+        }
+      } else {
+        this.reset();
+      }
+      return;
+    }
+  }
+
+  /**
+   * Save previous panel's focused element for later. Not necessary for view sections
+   */
+  private _savePrevElementState(prev: Region | undefined) {
+    const prevIsPanel = prev?.type === 'panel';
+    if (!prevIsPanel) {
+      return;
+    }
+    const prevPanelElement = getPanelElement(prev.id);
+    const isChildOfPanel = containsActiveElement(prevPanelElement);
+    if (!isChildOfPanel) {
+      return;
+    }
+    this._prevFocusedElements[prev.id] = document.activeElement;
+  }
+
+  private _onStateChange(current: State | undefined, prev: State | undefined) {
+    if (isEqual(current, prev)) {
+      return;
+    }
+
+    const gristDoc = this._getGristDoc();
+    const mouseEvent = current?.initiator?.type === 'mouse'
+      ? current.initiator.event
+      : undefined;
+
+    disableFocusLock();
+    removeFocusRings();
+    removeTabIndexes();
+    if (!mouseEvent) {
+      this._savePrevElementState(prev?.region);
+      if (prev?.region?.type === 'panel') {
+        blurPanelChild(prev.region);
+      }
+    }
+
+    const isPanel = current?.region?.type === 'panel';
+    const panelElement = isPanel && current.region?.id && getPanelElement((current.region as PanelRegion).id);
+
+    // if kb-focusing a panel: actually focus panel element
+    if (!mouseEvent && isPanel && panelElement && current.region) {
+      focusPanel(
+        current.region as PanelRegion,
+        this._prevFocusedElements[current.region.id as Panel] as HTMLElement | null,
+        gristDoc
+      );
+
+    // if clicking on a panel: just make sure view layout commands are disabled
+    } else if (mouseEvent && isPanel && panelElement && gristDoc) {
+      escapeViewLayout(gristDoc, !!(mouseEvent.target as Element)?.closest(`[${ATTRS.regionId}="right"]`));
+
+    // if clicking or kb-focusing a section: focus the section
+    } else if (current?.region?.type === 'section' && gristDoc) {
+      focusSection(current.region, gristDoc);
+    }
+
+    // if we reset the focus switch, clean all necessary state
+    if (current === undefined) {
+      if (gristDoc) {
+        focusViewLayout(gristDoc);
+      }
+      this._prevFocusedElements = {
+        left: null,
+        top: null,
+        right: null,
+        main: null,
+      };
+    }
+  }
+
+  private _toggleCreatorPanel() {
+    const current = this._state.get().region;
+    const gristDoc = this._getGristDoc();
+    if (current?.type === 'panel' && current.id === 'right') {
+      return this.focusRegion(gristDoc
+        ? {type: 'section', id: gristDoc.viewModel.activeSectionId()}
+        : {type: 'panel', id: 'main'},
+        {initiator: {type: 'cycle'}}
+      );
+    }
+    commands.allCommands.rightPanelOpen.run();
+    return this.focusRegion({type: 'panel', id: 'right'}, {initiator: {type: 'cycle'}});
+  }
+
+  private _canTabThroughMainRegion(use: UseCBOwner) {
```

Comment:
A useful trick is to accept the `use` parameter as `(use: UseCB = unwrap)` (with `import {unwrap, UseCB} from 'grainjs'`). Then in the function do `use(observable)`, and when the `use` parameter is omitted, it would be equivalent to `observable.get()`.

If you accept such a parameter for `_getGristDoc` and `isSpecialPage`, then you could reuse their logic here, and this function (if my understanding is correct) should just become
```
  private _canTabThroughMainRegion(use: UseCB = unwrap) {
    return !this._getGristDoc(use);
  }
```